src/Microsoft.Identity.Web.TokenAcquisition/TokenAcquisition.cs

```diff
@@ -745,35 +746,40 @@ private bool IsInvalidClientCertificateOrSignedAssertionError(MsalServiceExcepti
                 || exMsal.Message.Contains(Constants.CertificateHasBeenRevoked, StringComparison.OrdinalIgnoreCase)
                 || exMsal.Message.Contains(Constants.CertificateIsOutsideValidityWindow, StringComparison.OrdinalIgnoreCase));
 #else
-                (exMsal.Message.Contains(Constants.InvalidKeyError) 
-                || exMsal.Message.Contains(Constants.SignedAssertionInvalidTimeRange) 
+                (exMsal.Message.Contains(Constants.InvalidKeyError)
+                || exMsal.Message.Contains(Constants.SignedAssertionInvalidTimeRange)
                 || exMsal.Message.Contains(Constants.CertificateHasBeenRevoked)
                 || exMsal.Message.Contains(Constants.CertificateIsOutsideValidityWindow));
 #endif
-        }
-        
+        }       
+
+
         internal /* for testing */ async Task<IConfidentialClientApplication> GetOrBuildConfidentialClientApplicationAsync(
-           MergedOptions mergedOptions)
+             MergedOptions mergedOptions)
         {
-            if (!_applicationsByAuthorityClientId.TryGetValue(GetApplicationKey(mergedOptions), out IConfidentialClientApplication? application) || application == null)
+            // Use all credentials to compute a credential chain ID. Each individual ID should be unique.
+            string credentialId = string.Join("-", mergedOptions.ClientCredentials?.Select(c => c.Id) ?? Enumerable.Empty<string>());
```

Comment:
This can be very big
Also there are credentials that don't apply. But I guess this is ok.

Reply:
Discussed to leave as is.